### ISSUE TITLE
Add option to include partitioning columns in basename_template's filename

### ISSUE THREAD

**Issue by davlee1972** [apache/arrow#39447]
### Describe the enhancement requested

The default basename_template is “part-{i}.” + format.default_extname in pyarrow.dataset.write_dataset

https://arrow.apache.org/docs/python/generated/pyarrow.dataset.write_dataset.html

Can we add partitioning_columns in the template.. If I have additional partitioning columns like:
fx_currency
fx_year
fx_month

Could we support a basename_template like "fx_rates.{fx_currency}_{fx_year}_{fx_month}.part-{i}.parquet"?

This would ensure all filenames are unique even if they use hive or directory partitioning..

### Component(s)

C++, Python

**Comment by jorisvandenbossche** [comment]
Just a note that there exists a `FilenamePartitioning` (eg see third bullet point in https://arrow.apache.org/docs/python/generated/pyarrow.dataset.partitioning.html), that will give you file names with the partitioning values included. However, that's a custom filename-based partitioning scheme, and so you cannot combine that with a hive or directory partitioning (and so I assume not what you want in this case).

**Comment by davlee1972** [comment]
Correct. If filename partitioning could be used in conjunction with hive or directory partitioning that would work. Right now I’m using a hacked file visitor callback to rename files but that creates a conflict file behavior when deciding to overwrite, append or error existing files.

**Comment by davlee1972** [comment]
I managed to implement a custom partitioning class with a "scan" function which works for dataset read(s).. But dataset.write_data() only works with one of the three predefined "flavors".. 

There is a file_visitor() callback that can be used to rename the file, but it is very hacky to use especially if you want to replace existing files.

Also implementing a custom partitioning class that works in conjunction with hive partitioning is a bit awkward..

I think a proper implementation would be to split directory partitioning and file partitioning into separate configurations.

hive and directory partitioning would used for subdirectories.
filename partitioning with a filename template would be used for files.
filename_template = "fx_rates.{fx_currency}_{fx_year}_{fx_month}.part-{i}.parquet"
default would be filename_template = "{fx_currency}_{fx_year}_{fx_month}_" + basename_template. 

Also adding {i} should be optional and in practice should be used only if parquet files exceed 128 megs..

We have a lot of existing parquet files with naming conventions which do not conform to any of the current flavors available.

**Comment by thisisnic** [comment]
This issue has been marked as stale because it has had no activity in the past 2 years. Please remove the stale label or comment below, or this issue will be closed in 14 days.
